### PR TITLE
Correctly split train and test dataset

### PR DESCRIPTION
**Pull request by nstrelow**
* Load all images into one dataset and split dataset afterwards
* Config for train split size

### CODE REVIEW

File: multibuildingdetector/__main__.py [hunk 1]
```diff
@@ -36,12 +36,16 @@ def run(input, output, batch_size, iterator='SerialIterator', device=-1,
         chainer.cuda.get_device_from_id(device).use()
         model.to_gpu()
 
-    train = TransformDataset(
-        xmldataset.XMLDataset(input),
+    dataset = xmldataset.XMLDataset(input)
```
Comment (fapaul) [inline multibuildingdetector/__main__.py]: We should not implement the train test split logic in `__main__.py`. The dataset should handle this.

File: multibuildingdetector/__main__.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def load_config(path):
         return yaml.load(f)
 
 
-def run(input, output, batch_size, iterator='SerialIterator', device=-1,
+def run(input, output, batch_size, train_split, iterator='SerialIterator', device=-1,
```
Comment (fapaul) [inline multibuildingdetector/__main__.py]: Please add a default value eg. 80/20 for the split and this line is too long >79.

File: multibuildingdetector/readers/xmldataset.py [hunk 3]
```diff
@@ -36,10 +36,7 @@ def get_example(self, i):
         Returns:
             tuple of an image and bounding boxes and label
         """
-        if self._split == 'train':
-            id_ = self._img_ids[i * 2]
-        else:
-            id_ = self._img_ids[i * 2 + 1]
+        id_ = self._img_ids[i]
```
Comment (fapaul) [inline multibuildingdetector/readers/xmldataset.py]: I think due to this change the `__len__` attribute is not correct anymore. Probably @shorschig knows more about why this is needed.

**Review by fapaul — CHANGES_REQUESTED** [review]

**Comment by fapaul** [comment]
It would be great to also find a very representative split when looking through the labels. Currently it's likely that we generate biased sets.

File: multibuildingdetector/readers/xmldataset.py (hunk 3 above)
Comment (nstrelow) [reply multibuildingdetector/readers/xmldataset.py]: You are right, the len attribute is broken. Will fix that

This was used to split the dataset by odd and even index numbers, which always results into a 50:50 split.

File: multibuildingdetector/__main__.py (hunk 1 above)
Comment (nstrelow) [reply multibuildingdetector/__main__.py]: The dataset class should only contain and serve data in the format we want it to.
Everything else, like what it is used for, should be handled separately, like train/test split before initializing the iterators and start training.

**Comment by nstrelow** [comment]
I agree, we would need to separate train/test on our own to have all labels in the same train/test split balance.

We probably should do some random shuffle on each label and then divide into the train/test set.

Chainers MNIST example loads train/test from two different folders and creates two datasets.
See: https://github.com/chainer/chainer/blob/3df24dda17242d28ced4b26f4c9c897be7d15099/chainer/datasets/mnist.py#L48

File: multibuildingdetector/__main__.py [hunk 4]
```diff
@@ -36,12 +36,13 @@ def run(input, output, batch_size, iterator='SerialIterator', device=-1,
         chainer.cuda.get_device_from_id(device).use()
         model.to_gpu()
 
-    train = TransformDataset(
-        xmldataset.XMLDataset(input),
+    train, test = reader.load_train_test_set(input_dir, 'xml', 0.8)
```
Comment (sfriedo) [inline multibuildingdetector/__main__.py]: We should not look at the xml files only, but at the JPG files. If there are no bboxes on a picture, no xml exists, but we need them for the 'none' category (unless proven otherwise).

File: multibuildingdetector/readers/reader.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+from os import listdir
+from os.path import isfile, join
+from sklearn.model_selection import train_test_split
+
+from xmldataset import XMLDataset
+
+
+def load_train_test_set(data_dir, annotationType='xml', train_size=0.8):
```
Comment (sfriedo) [inline multibuildingdetector/readers/reader.py]: see above, jpg not xml

**Review by sfriedo — CHANGES_REQUESTED** [review]

File: multibuildingdetector/__main__.py (hunk 4 above)
Comment (nstrelow) [reply multibuildingdetector/__main__.py]: Ok good point, but what if our files are mixed xml and json?

File: config/cpu_default.yml [hunk 6]
```diff
@@ -1,6 +1,7 @@
-input: ../labeled_images/
+input_dir: ../labeled_images/
```
Comment (fapaul) [inline config/cpu_default.yml:1]: Adjust `README.md` with new options

File: multibuildingdetector/__main__.py (hunk 4 above)
Comment (fapaul) [reply multibuildingdetector/__main__.py]: I think we should, let the user configure the specific reader class and we just provide the interface with the `load_train_test` method. This way we do not have to care about the file endings etc.

File: multibuildingdetector/readers/reader.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+from .xmldataset import XMLDataset
+
+
+def load_train_test_set(data_dir, annotationType='xml', train_size=0.8):
+    img_id_paths = [join(data_dir, ''.join(f.split('.')[:-1])) for f in
+                    listdir(data_dir) if isfile(join(data_dir, f)) and f.split('.')[-1] == annotationType]
+
+    train_id_paths, test_id_paths = train_test_split(img_id_paths, train_size=train_size)
+    print("Split: {0}, Train size: {1}, Test size: {2}".format(train_size, len(train_id_paths), len(test_id_paths)))
+    return XMLDataset(train_id_paths), XMLDataset(test_id_paths)
```
Comment (fapaul) [inline multibuildingdetector/readers/reader.py]: As mentioned above we should allow multiple parser classes configured by the user.

File: multibuildingdetector/readers/reader.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+from .xmldataset import XMLDataset
+
+
+def load_train_test_set(data_dir, annotationType='xml', train_size=0.8):
+    img_id_paths = [join(data_dir, ''.join(f.split('.')[:-1])) for f in
+                    listdir(data_dir) if isfile(join(data_dir, f)) and f.split('.')[-1] == annotationType]
+
+    train_id_paths, test_id_paths = train_test_split(img_id_paths, train_size=train_size)
+    print("Split: {0}, Train size: {1}, Test size: {2}".format(train_size, len(train_id_paths), len(test_id_paths)))
```
Comment (fapaul) [inline multibuildingdetector/readers/reader.py]: Multiple lines are too long. Please install a pep8 linter 👍

File: multibuildingdetector/readers/reader.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+from os import listdir
+from os.path import isfile, join
+from sklearn.model_selection import train_test_split
+
+from .xmldataset import XMLDataset
+
+
+def load_train_test_set(data_dir, annotationType='xml', train_size=0.8):
+    img_id_paths = [join(data_dir, ''.join(f.split('.')[:-1])) for f in
```
Comment (fapaul) [inline multibuildingdetector/readers/reader.py]: Is `img_id_paths` == `img_paths`?

**Review by fapaul — CHANGES_REQUESTED** [review]

File: multibuildingdetector/__main__.py (hunk 4 above)
Comment (nstrelow) [reply multibuildingdetector/__main__.py]: This can be done in a separate PR

**Review by sfriedo — APPROVED** [review]

**Review by fapaul — APPROVED** [review]
